### PR TITLE
monasca: Register logs endpoints in keystone

### PR DESCRIPTION
**Pull request by samirjorina**
This PR needs back porting to stable/4.0, unfortunately I don't have permissions (?) to add the appropriate label :(

### CODE REVIEW

**Comment by jgrassler** [comment]
Good and bad news about this. The log API tests _do_ still fail...

```
ostestr '.*monasca.*' 2>&1 | tee tempest.log
[...]
{0} monasca_log_api_tempest.tests.log_api.test_constraints.TestLogApiConstraints.test_should_reject_too_big_message [1.982807s] ... FAILED
{0} monasca_log_api_tempest.tests.log_api.test_constraints.TestLogApiConstraints.test_should_reject_too_big_message_multiline [3.248677s] ... FAILED
{1} monasca_log_api_tempest.tests.log_api.test_single.TestSingleLog.test_big_message [9.472205s] ... FAILED
{1} monasca_log_api_tempest.tests.log_api.test_single.TestSingleLog.test_big_message_multiline [1.028245s] ... FAILED
{1} monasca_log_api_tempest.tests.log_api.test_single.TestSingleLog.test_medium_message [0.778157s] ... FAILED
{1} monasca_log_api_tempest.tests.log_api.test_single.TestSingleLog.test_medium_message_multiline [0.831534s] ... FAILED
{1} monasca_log_api_tempest.tests.log_api.test_single.TestSingleLog.test_send_header_application_type [0.903323s] ... FAILED
{1} monasca_log_api_tempest.tests.log_api.test_single.TestSingleLog.test_send_header_dimensions [0.790627s] ... FAILED
{1} monasca_log_api_tempest.tests.log_api.test_single.TestSingleLog.test_small_message [0.814206s] ... FAILED
{1} monasca_log_api_tempest.tests.log_api.test_single.TestSingleLog.test_small_message_multiline [1.416836s] ... FAILED
```

...but no longer due to missing endpoints but unexpected responses from the correct log API. Here's an example:

```
monasca_log_api_tempest.tests.log_api.test_constraints.TestLogApiConstraints.test_should_reject_too_big_message_multiline[gate]
-------------------------------------------------------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/monasca_log_api_tempest/tests/log_api/test_constraints.py", line 99, in test_should_reject_too_big_message_multiline
        self.assertTrue(False, 'API should respond with 413')
      File "/usr/lib/python2.7/site-packages/unittest2/case.py", line 702, in assertTrue
        raise self.failureException(msg)
    AssertionError: False is not true : API should respond with 413
    

Captured pythonlogging:
~~~~~~~~~~~~~~~~~~~~~~~
    2017-06-21 16:00:28,906 2891 INFO     [urllib3.connectionpool] Resetting dropped connection: 192.168.232.5
    2017-06-21 16:00:29,100 2891 DEBUG    [urllib3.connectionpool] "POST /v2.0/log/single HTTP/1.1" 204 0
```

The URL being queried is the logs_v2 public URL:

```
root@d52-54-77-77-01-01:/var/lib/openstack-tempest-test # openstack endpoint list | grep logs_v2 | grep public
| 97a4531710cd40a081deda90ae457510 | CustomRegion | logs_v2      | logs_v2        | True    | public    | http://192.168.232.5:5607/v2.0
```

So at least the part this pull request is about works.

**Comment by samirjorina** [comment]
Most of the failures is related to misconfiguration of log-api and/or tempest tests.
One of them (test_should_reject_if_body_is_empty) seems to be broken, I'll try to fix it.
Anyway, the failing test has nothing to do with this PR :)

File: chef/cookbooks/monasca/libraries/helper.rb [hunk 1]
```diff
@@ -73,6 +73,75 @@ def self.log_api_public_url(node)
...
+  def self.log_api_internal_url(node)
+    host = get_host_for_monitoring_url(node)
+    # SSL is not supported at this moment
+    # protocol = node[:monasca][:log_api][:ssl] ? "https" : "http"
+    protocol = "http"
+    port = node[:monasca][:log_api][:bind_port]
+    "#{protocol}://#{host}:#{port}/v3.0"
+  end
+
+  def self.log_api_v2_public_url(node)
```
Comment (toabctl) [inline chef/cookbooks/monasca/libraries/helper.rb]: Can we remove this method and just use self.logs_search_public_url and append the /v2.0 ?

File: chef/cookbooks/monasca/libraries/helper.rb [hunk 2]
```diff
@@ -73,6 +73,75 @@ def self.log_api_public_url(node)
...
+    host = monasca_public_host(node)
+    # SSL is not supported at this moment
+    # protocol = node[:monasca][:log_api][:ssl] ? "https" : "http"
+    protocol = "http"
+    port = node[:monasca][:log_api][:bind_port]
+    "#{protocol}://#{host}:#{port}/v2.0"
+  end
+
+  def self.log_api_v2_admin_url(node)
```
Comment (toabctl) [inline chef/cookbooks/monasca/libraries/helper.rb]: same here

File: chef/cookbooks/monasca/libraries/helper.rb [hunk 3]
```diff
@@ -73,6 +73,75 @@ def self.log_api_public_url(node)
...
+    host = monasca_admin_host(node)
+    # SSL is not supported at this moment
+    # protocol = node[:monasca][:log_api][:ssl] ? "https" : "http"
+    protocol = "http"
+    port = node[:monasca][:log_api][:bind_port]
+    "#{protocol}://#{host}:#{port}/v2.0"
+  end
+
+  def self.log_api_v2_internal_url(node)
```
Comment (toabctl) [inline chef/cookbooks/monasca/libraries/helper.rb]: same here

File: chef/cookbooks/monasca/recipes/server.rb [hunk 4]
```diff
@@ -80,6 +80,84 @@
...
+
+keystone_register "register logs_v2 endpoint" do
+  protocol keystone_settings["protocol"]
+  insecure keystone_settings["insecure"]
+  host keystone_settings["internal_url_host"]
+  port keystone_settings["admin_port"]
+  auth register_auth_hash
+  endpoint_service "logs_v2"
+  endpoint_region keystone_settings["endpoint_region"]
+  endpoint_publicURL MonascaHelper.log_api_v2_public_url(node)
```
Comment (toabctl) [inline chef/cookbooks/monasca/recipes/server.rb]: as mentioned, use log_api_public_url and append the v2.0

File: chef/cookbooks/monasca/libraries/helper.rb (hunk 1 above)
Comment (samirjorina) [inline chef/cookbooks/monasca/libraries/helper.rb]: `logs_search_public_url` (kibana) uses different port (5601) than `log_api_v2_public_url` (monasca-log-api - 5607)

Not sure how to make it nicer - maybe I could merge log_api and log_api_v2 into single method returning unversioned endpoint and then append version in the keystone registration?

Comment (toabctl) [inline chef/cookbooks/monasca/libraries/helper.rb]: @samirjorina yeah. That's what I meant. Sorry for the confusion with the method names.

**Pull request merged**